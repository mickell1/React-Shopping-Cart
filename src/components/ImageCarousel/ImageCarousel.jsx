import React, { Component } from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

import styles from './ImageCarousel.scss'

class ImageCarousel extends Component {
  constructor(props) {
    super(props)
    this.state = {
      activeImage: 1,
      imagesCount: props.images.length, // eslint-disable-line react/no-unused-state
    }
  }

  componentDidMount() {
    this.fadeImages()
  }

  componentWillUnmount() {
    this.clear()
  }

  clear = () => {
    clearInterval(this.timer)
  }

  fadeImages = () => {
    this.timer = setInterval(() => {
      this.setState((state) => {
        const activeImage = state.activeImage < state.imagesCount
          ? state.activeImage
          : 0
        return {
          activeImage: activeImage + 1,
          nextImage: activeImage + 2,
        }
      })
    }, 5000)
  }

  render() {
    const {
      images,
      imageAltText,
    } = this.props
    return (
      <div className={styles['image-slider']} onMouseLeave={this.fadeImages} onMouseEnter={this.clear}>
        {images.map((image, index) => {
          const imageClasses = classnames({
            [styles['image-container']]: true,
            [styles.active]: this.state.activeImage === index + 1,
            [styles.next]: this.state.nextImage === index + 1,
          })
          return (
            <div className={imageClasses} key={index}>
              <img src={image} alt={imageAltText} />
            </div>
          )
        })}
        <div className={styles['dots-container']}>
          {images.map((image, index) => {
            const dotClasses = classnames({
              [styles.dot]: true,
              [styles['dot-active']]: this.state.activeImage === index + 1,
            })
            return (
              <div className={dotClasses} key={`dot-${index}`} />
            )
          })}
        </div>
      </div>

    )
  }
}

ImageCarousel.propTypes = {
  images: PropTypes.arrayOf(PropTypes.string),
  imageAltText: PropTypes.string,
}

export default ImageCarousel
