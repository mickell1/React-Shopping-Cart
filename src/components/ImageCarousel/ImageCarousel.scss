@import "../../variables.scss";

.image-slider {
  width: 100%;
  position: relative;
}

.image-container {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  z-index: 1;
  background-size: cover;
  background-repeat: no-repeat;
  background-position: center center;
  opacity: 0;
  transition: opacity 1s ease;
}

.image-container.active {
  z-index: 3;
  opacity: 1;
  position: relative;
}

.image-container.next {
  z-index: 2;
}

.dots-container {
  display: flex;
  width: 100%;
  margin-top: 30px;
  justify-content: center;
}

.dot {
  width: 10px;
  height: 10px;
  margin-right: 10px;
  border-radius: 50%;
  background-color: #ccc;
  transition: background-color .3s ease;

  &.dot-active {
    background-color: $color-black;
  }
}
