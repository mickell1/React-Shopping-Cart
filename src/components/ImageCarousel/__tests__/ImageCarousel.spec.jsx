import React from 'react'
import { shallow, mount } from 'enzyme'
import ImageCarousel from '../ImageCarousel'

const defaultProps = {
  images: [
    'test',
    'test2',
  ],
  imageAltText: 'Test Text',
  componentDidMount: jest.fn(),
  componentWillUnmount: jest.fn(),
  clear: jest.fn(),
  fadeImages: jest.fn(),
}

let wrapper;

beforeEach(()=>{
  wrapper = shallow(<ImageCarousel {...defaultProps} /> )  
})

describe('ImageCarousel', () => {
  it('should render the ImageCarousel Component correctly', () => {   
    expect(wrapper).toMatchSnapshot();
  });

  it('should render props', () => {
    wrapper = mount(<ImageCarousel {...defaultProps} />)
    expect(wrapper.props().images).toEqual(['test', 'test2'])
    expect(wrapper.props().imageAltText).toEqual('Test Text')
  })
});
