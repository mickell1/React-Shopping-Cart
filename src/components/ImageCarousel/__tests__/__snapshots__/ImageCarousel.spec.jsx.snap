// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageCarousel should render the ImageCarousel Component correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ImageCarousel
    clear={[MockFunction]}
    componentDidMount={[MockFunction]}
    componentWillUnmount={[MockFunction]}
    fadeImages={[MockFunction]}
    imageAltText="Test Text"
    images={
      Array [
        "test",
        "test2",
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        Array [
          <div
            className=""
          >
            <img
              alt="Test Text"
              src="test"
            />
          </div>,
          <div
            className=""
          >
            <img
              alt="Test Text"
              src="test2"
            />
          </div>,
        ],
        <div>
          <div
            className="undefined"
          />
          <div
            className=""
          />
        </div>,
      ],
      "className": undefined,
      "onMouseEnter": [Function],
      "onMouseLeave": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "0",
        "nodeType": "host",
        "props": Object {
          "children": <img
            alt="Test Text"
            src="test"
          />,
          "className": "",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "alt": "Test Text",
            "src": "test",
          },
          "ref": null,
          "rendered": null,
          "type": "img",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "host",
        "props": Object {
          "children": <img
            alt="Test Text"
            src="test2"
          />,
          "className": "",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "alt": "Test Text",
            "src": "test2",
          },
          "ref": null,
          "rendered": null,
          "type": "img",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="undefined"
            />,
            <div
              className=""
            />,
          ],
          "className": undefined,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "dot-0",
            "nodeType": "host",
            "props": Object {
              "className": "undefined",
            },
            "ref": null,
            "rendered": null,
            "type": "div",
          },
          Object {
            "instance": null,
            "key": "dot-1",
            "nodeType": "host",
            "props": Object {
              "className": "",
            },
            "ref": null,
            "rendered": null,
            "type": "div",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          Array [
            <div
              className=""
            >
              <img
                alt="Test Text"
                src="test"
              />
            </div>,
            <div
              className=""
            >
              <img
                alt="Test Text"
                src="test2"
              />
            </div>,
          ],
          <div>
            <div
              className="undefined"
            />
            <div
              className=""
            />
          </div>,
        ],
        "className": undefined,
        "onMouseEnter": [Function],
        "onMouseLeave": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "0",
          "nodeType": "host",
          "props": Object {
            "children": <img
              alt="Test Text"
              src="test"
            />,
            "className": "",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "alt": "Test Text",
              "src": "test",
            },
            "ref": null,
            "rendered": null,
            "type": "img",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "host",
          "props": Object {
            "children": <img
              alt="Test Text"
              src="test2"
            />,
            "className": "",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "alt": "Test Text",
              "src": "test2",
            },
            "ref": null,
            "rendered": null,
            "type": "img",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="undefined"
              />,
              <div
                className=""
              />,
            ],
            "className": undefined,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "dot-0",
              "nodeType": "host",
              "props": Object {
                "className": "undefined",
              },
              "ref": null,
              "rendered": null,
              "type": "div",
            },
            Object {
              "instance": null,
              "key": "dot-1",
              "nodeType": "host",
              "props": Object {
                "className": "",
              },
              "ref": null,
              "rendered": null,
              "type": "div",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
